bin/resource_management/resource-broken-links-csv.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[ ]:
+
+
+import pandas as pd
+import urllib
+import sys
+from sys import stderr, stdout, argv
+from multiprocessing import Process
+
+
+# In[ ]:
+
+
+def run_cpu_tasks_in_parallel(tasks):
+    """
+    Manages processes
+    tasks - list of methods to be run
+    """
+    running_tasks = [Process(target=task) for task in tasks]
+    for running_task in running_tasks:
+        running_task.start()
+    for running_task in running_tasks:
+        running_task.join()
+        
+def main(filename,processes=4):
+    """
+    Scan url's of resources from dataset metadata obtained from https://open.canada.ca/data/en
+    CKAN API - ckanapi dump datasets --all -O datasets.jsonl.gz -z -p 4 -r http://localhost
+    
+    filename - destination of jsonl/jl file 
+    processes - default processes=4. Number of batches and processes to run. Will export 1 file per process.
+    
+    Exports broken links found in metadata
+    """
+    try:
+        df=pd.read_json(filename,lines=True)
+    except:
+        print('Error: .jsonl/.jl file expected')
+        return False
+    file_length = len(df)
+    batchsize=round(file_length/processes)
+    tasks=[]
+    for i in range(processes):
+        lower = i*batchsize
+        upper = (i+1)*batchsize
+        if(i==processes-1):
+            upper=file_length
+        tasks.append(verify_url(lower,upper,i,df))
+    run_cpu_tasks_in_parallel(tasks)
+
+
+# In[ ]:
+
+
+def verify_url (lower,upper,batch_number,df):
+    """
+    Scan url's of resources of current batch from dataset metadata obtained from https://open.canada.ca/data/en
+    
+    lower - starting index of batch
+    upper - final index of batch
+    batch_number - current batch number
+    df - copy of metadata dataframe
+    
+    exports CSV of broken links from current batch - 'od-broken-links_batch{0}.csv'.format(batch_number)
+    """
+    broken_links=[]
+    for i in range(lower,upper):
+        resources = df["resources"][i]
+        error='no-code'
+        for z in range(len(resources)):
+            flag=0
+            if("url" in resources[z]):
+                name=resources[z].get("name")
+                url=resources[z].get("url")
+                try:
+                    c = urllib.request.urlopen(url)
```

Comment:
the requests library handles http more reliably than urllib and is even easier to use

Reply:
I started out using requests but the info I received from getting headers was inconsistent. Maybe I was implementing it the wrong way?

Reply:
interesting. I thought the requests docs were pretty good. What didn't work about getting the headers?

Reply:
Often the http code would be None, and requests also does not work with FTP urls